File: 3601-find-the-k-th-character-in-string-game-ii/3601-find-the-k-th-character-in-string-game-ii.py
class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        ## so how many operations aplied to a?
        k = k-1 # first position is 0
        a = bin(k)[2:] # turn into binary, remove bx header
        a = a[::-1] # reverse 
        a = a + "0"*(len(operations)-len(a)) #add zeros
        
        counter= 0 # merge both lists
        for a,b in zip(a,operations):
            if a=="1" and b==1:
                counter+=1
        return chr(ord('a')+counter%26) #26 is the size of the alphabet
       
                

File: 3601-find-the-k-th-character-in-string-game-ii/3601-find-the-k-th-character-in-string-game-ii.cpp
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        int n = operations.size();
        vector<long long> len(n + 1, 1);
        for (int i = 0; i < n; ++i) {
            len[i + 1] = min<long long>(len[i] << 1, (long long)2e18);
        }
        int shift = 0;
        for (int i = n - 1; i >= 0; --i) {
            long long half = len[i];
            if (k > half) {
                k -= half;
                if (operations[i]) shift = (shift + 1) % 26;
            }
        }
        return char('a' + shift);
    }
};